File: realtime/android/webrtc/app/src/test/java/ai/inworld/voiceagent/ClientEventEncodingTest.kt
package ai.inworld.voiceagent

import ai.inworld.voiceagent.realtime.BackchannelConfig
import ai.inworld.voiceagent.realtime.ResponsivenessConfig
import ai.inworld.voiceagent.realtime.SessionConfig
import ai.inworld.voiceagent.realtime.TurnDetectionConfig
import ai.inworld.voiceagent.realtime.WebSearchConfig
import ai.inworld.voiceagent.realtime.events.ConversationItemCreateEvent
import ai.inworld.voiceagent.realtime.events.EventJson
import ai.inworld.voiceagent.realtime.events.ResponseCancelEvent
import ai.inworld.voiceagent.realtime.events.ResponseCreateEvent
import ai.inworld.voiceagent.realtime.events.SessionUpdateEvent
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class ClientEventEncodingTest {
    private inline fun <reified T> encodeToObject(event: T): JsonObject =
        EventJson.parseToJsonElement(EventJson.encodeToString(event)).jsonObject

    private fun baseConfig() = SessionConfig(
        model = "openai/gpt-4o-mini",
        instructions = "Be brief.",
        ttsModel = "inworld-tts-2",
        voice = "Clive",
        greetingPrompt = "Say hello.",
    )

    private fun JsonObject.obj(key: String): JsonObject = getValue(key).jsonObject
    private fun JsonObject.str(key: String): String? = get(key)?.jsonPrimitive?.contentOrNull
    private fun JsonObject.int(key: String): Int? = get(key)?.jsonPrimitive?.intOrNull
    private fun JsonObject.double(key: String): Double? = get(key)?.jsonPrimitive?.doubleOrNull
    private fun JsonObject.bool(key: String): Boolean? = get(key)?.jsonPrimitive?.booleanOrNull

    @Test
    fun sessionUpdateDefaultShape() {
        val json = encodeToObject(SessionUpdateEvent.from(baseConfig()))

        assertEquals("session.update", json.str("type"))
        val session = json.obj("session")
        assertEquals("realtime", session.str("type"))
        assertEquals("openai/gpt-4o-mini", session.str("model"))
        assertEquals("Be brief.", session.str("instructions"))
        assertEquals(
            listOf("audio", "text"),
            session.getValue("output_modalities").jsonArray.map { it.jsonPrimitive.contentOrNull },
        )
        assertNull(session["temperature"])
        assertNull(session["max_output_tokens"])
        assertNull(session["tools"])

        val input = session.obj("audio").obj("input")
        assertNull(input["transcription"])
        assertNull(input["noise_reduction"])

        val turnDetection = input.obj("turn_detection")
        assertEquals("semantic_vad", turnDetection.str("type"))
        assertEquals("high", turnDetection.str("eagerness"))
        assertEquals(true, turnDetection.bool("create_response"))
        assertEquals(true, turnDetection.bool("interrupt_response"))
        assertNull(turnDetection["threshold"])

        val output = session.obj("audio").obj("output")
        assertEquals("inworld-tts-2", output.str("model"))
        assertEquals("Clive", output.str("voice"))
        assertNull(output["speed"])
    }

    @Test
    fun sessionUpdateWithWebSearchEncodesToolWithCamelCaseProviderData() {
        val config = baseConfig().copy(webSearch = WebSearchConfig())
        val raw = EventJson.encodeToString(SessionUpdateEvent.from(config))
        assertTrue(raw.contains("\"providerData\""))
        assertFalse(raw.contains("\"provider_data\""))

        val session = encodeToObject(SessionUpdateEvent.from(config)).obj("session")
        val tools = session.getValue("tools").jsonArray
        assertEquals(1, tools.size)
        val tool = tools[0].jsonObject
        assertEquals("web_search", tool.str("type"))
        val pd = tool.obj("providerData")
        assertEquals("google", pd.str("engine"))
        assertEquals(3, pd.int("max_results"))
        assertEquals(1, pd.int("max_steps"))
    }

    @Test
    fun sessionUpdateWithAllOptions() {
        val config = baseConfig().copy(
            temperature = 0.9,
            maxOutputTokens = 1024,
            speechSpeed = 1.25,
            transcriptionModel = "inworld/inworld-stt-1",
            transcriptionLanguage = "en",
            noiseReduction = "near_field",
            turnDetection = TurnDetectionConfig(eagerness = "low"),
            createResponse = false,
            interruptResponse = false,
        )

        val session = encodeToObject(SessionUpdateEvent.from(config)).obj("session")
        assertEquals(0.9, session.double("temperature")!!, 0.0)
        assertEquals(1024, session.int("max_output_tokens"))

        val input = session.obj("audio").obj("input")

        val transcription = input.obj("transcription")
        assertEquals("inworld/inworld-stt-1", transcription.str("model"))
        assertEquals("en", transcription.str("language"))

        assertEquals("near_field", input.obj("noise_reduction").str("type"))

        val turnDetection = input.obj("turn_detection")
        assertEquals("semantic_vad", turnDetection.str("type"))
        assertEquals("low", turnDetection.str("eagerness"))
        assertEquals(false, turnDetection.bool("create_response"))
        assertEquals(false, turnDetection.bool("interrupt_response"))

        assertEquals(1.25, session.obj("audio").obj("output").double("speed")!!, 0.0)
    }

    @Test
    fun providerDataOmittedByDefault() {
        val session = encodeToObject(SessionUpdateEvent.from(baseConfig())).obj("session")
        assertNull(session["providerData"])
        assertNull(session["provider_data"])
    }

    @Test
    fun providerDataKeyIsCamelCaseWhileInnerKeysAreSnakeCase() {
        val config = baseConfig().copy(
            backchannel = BackchannelConfig(
                maxPerTurn = 3, minGapMs = 4000, minSpeechMs = 800, volumeGain = 0.6,
                deciderKind = "llm", ruleFireProbability = 1.0,
            ),
        )
        // Assert the raw JSON string, since the camelCase exception is the whole point.
        val raw = EventJson.encodeToString(SessionUpdateEvent.from(config))
        assertTrue("providerData must stay camelCase", raw.contains("\"providerData\""))
        assertFalse("must not be snake_cased", raw.contains("\"provider_data\""))
        assertTrue("inner keys stay snake_case", raw.contains("\"max_per_turn\""))
        assertTrue(raw.contains("\"output_modalities\""))
    }

    @Test
    fun backchannelAndResponsivenessEncoding() {
        val config = baseConfig().copy(
            backchannel = BackchannelConfig(
                maxPerTurn = 2, minGapMs = 3000, minSpeechMs = 600, volumeGain = 0.5,
                deciderKind = "rule", ruleFireProbability = 0.25,
            ),
            responsiveness = ResponsivenessConfig(
                initialWaitTimeoutMs = 900, maxInitialPerTurn = 1, minFillerGapMs = 6000,
                maxTokens = 10, enableOnFirstReply = true,
            ),
        )

        val providerData = encodeToObject(SessionUpdateEvent.from(config)).obj("session").obj("providerData")

        val backchannel = providerData.obj("backchannel")
        assertEquals(true, backchannel.bool("enabled"))
        assertEquals(2, backchannel.int("max_per_turn"))
        assertEquals(3000, backchannel.int("min_gap_ms"))
        assertEquals(600, backchannel.int("min_speech_ms"))
        assertEquals(0.5, backchannel.double("volume_gain")!!, 0.0)
        assertEquals("rule", backchannel.str("decider_kind"))
        assertEquals(0.25, backchannel.double("rule_fire_probability")!!, 0.0)

        val responsiveness = providerData.obj("responsiveness")
        assertEquals(true, responsiveness.bool("enabled"))
        assertEquals(900, responsiveness.int("initial_wait_timeout_ms"))
        assertEquals(1, responsiveness.int("max_initial_per_turn"))
        assertEquals(6000, responsiveness.int("min_filler_gap_ms"))
        assertEquals(10, responsiveness.int("max_tokens"))
        assertEquals(true, responsiveness.bool("enable_filler_on_first_assistant_reply"))
    }

    @Test
    fun providerDataOnlyBackchannel() {
        val config = baseConfig().copy(
            backchannel = BackchannelConfig(
                maxPerTurn = 3, minGapMs = 4000, minSpeechMs = 800, volumeGain = 0.6,
                deciderKind = "llm", ruleFireProbability = 0.5,
            ),
        )
        val providerData = encodeToObject(SessionUpdateEvent.from(config)).obj("session").obj("providerData")
        // rule_fire_probability is omitted for the llm decider.
        assertNull(providerData.obj("backchannel")["rule_fire_probability"])
        assertNull(providerData["responsiveness"])
    }

    @Test
    fun conversationItemCreateShape() {
        val json = encodeToObject(ConversationItemCreateEvent.userText("Say hello."))
        assertEquals("conversation.item.create", json.str("type"))
        val item = json.obj("item")
        assertEquals("message", item.str("type"))
        assertEquals("user", item.str("role"))
        val content = item.getValue("content").jsonArray
        assertEquals(1, content.size)
        assertEquals("input_text", content[0].jsonObject.str("type"))
        assertEquals("Say hello.", content[0].jsonObject.str("text"))
    }

    @Test
    fun responseCreateAndCancel() {
        assertEquals("response.create", encodeToObject(ResponseCreateEvent()).str("type"))
        assertEquals("response.cancel", encodeToObject(ResponseCancelEvent()).str("type"))
    }
}

File: realtime/flutter/webrtc/test/client_event_encoding_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:inworld_voice_agent/realtime/events/client_events.dart';
import 'package:inworld_voice_agent/realtime/session_config.dart';

SessionConfig baseConfig({
  WebSearchConfig? webSearch,
  BackchannelConfig? backchannel,
  ResponsivenessConfig? responsiveness,
  double? speechSpeed,
  double? temperature,
  int? maxOutputTokens,
  String? transcriptionModel,
  String? transcriptionLanguage,
  String? noiseReduction,
}) =>
    SessionConfig(
      model: 'openai/gpt-4o-mini',
      instructions: 'Be brief.',
      ttsModel: 'inworld-tts-2',
      voice: 'Clive',
      greetingPrompt: 'Say hello.',
      webSearch: webSearch,
      backchannel: backchannel,
      responsiveness: responsiveness,
      speechSpeed: speechSpeed,
      temperature: temperature,
      maxOutputTokens: maxOutputTokens,
      transcriptionModel: transcriptionModel,
      transcriptionLanguage: transcriptionLanguage,
      noiseReduction: noiseReduction,
    );

void main() {
  group('sessionUpdateEvent', () {
    test('default shape', () {
      final json = sessionUpdateEvent(baseConfig());
      expect(json['type'], 'session.update');

      final session = json['session'] as Map<String, dynamic>;
      expect(session['type'], 'realtime');
      expect(session['model'], 'openai/gpt-4o-mini');
      expect(session['instructions'], 'Be brief.');
      expect(session['output_modalities'], ['audio', 'text']);
      expect(session.containsKey('temperature'), isFalse);
      expect(session.containsKey('max_output_tokens'), isFalse);
      expect(session.containsKey('tools'), isFalse);
      expect(session.containsKey('providerData'), isFalse);

      final input = (session['audio'] as Map)['input'] as Map;
      expect(input.containsKey('transcription'), isFalse);
      expect(input.containsKey('noise_reduction'), isFalse);

      final td = input['turn_detection'] as Map;
      expect(td['type'], 'semantic_vad');
      expect(td['eagerness'], 'high');
      expect(td['create_response'], true);
      expect(td['interrupt_response'], true);

      final output = (session['audio'] as Map)['output'] as Map;
      expect(output['model'], 'inworld-tts-2');
      expect(output['voice'], 'Clive');
      expect(output.containsKey('speed'), isFalse);
    });

    test('web search encodes a tool with camelCase providerData', () {
      final config = baseConfig(webSearch: const WebSearchConfig());
      final raw = jsonEncode(sessionUpdateEvent(config));
      expect(raw.contains('"providerData"'), isTrue);
      expect(raw.contains('"provider_data"'), isFalse);

      final session = sessionUpdateEvent(config)['session'] as Map<String, dynamic>;
      final tools = session['tools'] as List;
      expect(tools.length, 1);
      final tool = tools.first as Map;
      expect(tool['type'], 'web_search');
      final pd = tool['providerData'] as Map;
      expect(pd['engine'], 'google');
      expect(pd['max_results'], 3);
      expect(pd['max_steps'], 1);
    });

    test('providerData key stays camelCase while inner keys are snake_case', () {
      final config = baseConfig(backchannel: const BackchannelConfig());
      final raw = jsonEncode(sessionUpdateEvent(config));
      expect(raw.contains('"providerData"'), isTrue);
      expect(raw.contains('"provider_data"'), isFalse);
      expect(raw.contains('"max_per_turn"'), isTrue);
      expect(raw.contains('"output_modalities"'), isTrue);
    });

    test('providerData omitted by default', () {
      final session = sessionUpdateEvent(baseConfig())['session'] as Map;
      expect(session.containsKey('providerData'), isFalse);
    });

    test('back-channel and responsiveness encoding', () {
      final config = baseConfig(
        backchannel: const BackchannelConfig(deciderKind: 'rule', ruleFireProbability: 0.25),
        responsiveness: const ResponsivenessConfig(enableOnFirstReply: true),
      );
      final session = sessionUpdateEvent(config)['session'] as Map;
      final pd = session['providerData'] as Map;

      final bc = pd['backchannel'] as Map;
      expect(bc['enabled'], true);
      expect(bc['max_per_turn'], 3);
      expect(bc['decider_kind'], 'rule');
      expect(bc['rule_fire_probability'], 0.25);

      final resp = pd['responsiveness'] as Map;
      expect(resp['enabled'], true);
      expect(resp['enable_filler_on_first_assistant_reply'], true);
    });

    test('rule_fire_probability omitted for llm decider', () {
      final config = baseConfig(backchannel: const BackchannelConfig());
      final session = sessionUpdateEvent(config)['session'] as Map;
      final bc = (session['providerData'] as Map)['backchannel'] as Map;
      expect(bc.containsKey('rule_fire_probability'), isFalse);
    });

    test('optional fields encode when set', () {
      final config = baseConfig(
        temperature: 0.9,
        maxOutputTokens: 1024,
        speechSpeed: 1.25,
        transcriptionModel: 'inworld/inworld-stt-1',
        transcriptionLanguage: 'en',
        noiseReduction: 'near_field',
      );
      final session = sessionUpdateEvent(config)['session'] as Map;
      expect(session['temperature'], 0.9);
      expect(session['max_output_tokens'], 1024);

      final input = (session['audio'] as Map)['input'] as Map;
      final transcription = input['transcription'] as Map;
      expect(transcription['model'], 'inworld/inworld-stt-1');
      expect(transcription['language'], 'en');
      expect((input['noise_reduction'] as Map)['type'], 'near_field');
      expect(((session['audio'] as Map)['output'] as Map)['speed'], 1.25);
    });
  });

  test('conversationItemCreateEvent shape', () {
    final json = conversationItemCreateEvent('Say hello.');
    expect(json['type'], 'conversation.item.create');
    final item = json['item'] as Map;
    expect(item['type'], 'message');
    expect(item['role'], 'user');
    final content = item['content'] as List;
    expect(content.length, 1);
    expect((content.first as Map)['type'], 'input_text');
    expect((content.first as Map)['text'], 'Say hello.');
  });

  test('response.create and response.cancel', () {
    expect(responseCreateEvent()['type'], 'response.create');
    expect(responseCancelEvent()['type'], 'response.cancel');
  });
}
